### PR TITLE
feat: backport handle zombie connection

### PR DESCRIPTION
**Pull request by legendhimslef**
**Please describe the changes this PR makes and why it should be merged:**
Backports the following pull requests to version 14: #7581 
**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/discord.js/src/util/ShardEvents.js [hunk 1]
```diff
@@ -7,4 +7,5 @@ module.exports = {
   Ready: 'ready',
   Resumed: 'resumed',
   AllReady: 'allReady',
+  ZombieConnection: 'zombieConnection',
```
Comment (almeidx) [inline packages/discord.js/src/util/ShardEvents.js:10]: This was not added to the typings

**Review by almeidx — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/client/websocket/WebSocketShard.js [hunk 2]
```diff
@@ -356,6 +374,19 @@ class WebSocketShard extends EventEmitter {
...
+    // Step 1: Null the connection object
+    this.connection = null;
+
+    // Step 2: Set the shard status to Disconnected
+    this.status = Status.Disconnected;
+
+    // Step 3: Cache the old sequence (use to attempt a resume)
+    if (this.sequence !== -1) this.closeSequence = this.sequence;
+
     this.status = Status.Disconnected;
```
Comment (muchnameless) [inline packages/discord.js/src/client/websocket/WebSocketShard.js:390]:
```suggestion
```
🧹

**Review by muchnameless — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Comment by iCrawl** [comment]
The needs a rebase.

**Comment by Monix362** [comment]
DiscordCanary_CvGumNjs2z

Are there any options to fix this connection? I had to disable the GUILD_PRESENCE intention to get my bot to reconnect to the gateway and function normally.

**Comment by legendhimslef** [comment]
> DiscordCanary_CvGumNjs2z
> 
> Are there any options to fix this connection? I had to disable the GUILD_PRESENCE intention to get my bot to reconnect to the gateway and function normally.

I fixed it but I guess I messed up fetching upstream. This pr now has 227 files changed ;-;
@Monix362 The issue was, `this.connection.terminate` successfully closed the connection which emitted the onClose event, and thus `this.connection` is now null.

**Comment by Monix362** [comment]
It doesn't work. My bot has 34 shards, of which 2 tried to connect to the gateway, and the rest simply hung up and no longer want to connect. Zombie connection is handled, only nothing changes, reconnect=hang=reconnect.

This happens when I have the GUILD_PRESENCE intention enabled, but without it it works as usual.

**Comment by legendhimslef** [comment]
> happens

can you show me the debug logs and have you implemented all the changes done in this pr in your code? Like all the 3 files as it is.
Also, my bot has around 24 shards. Only one shard appears to get into a zombie connection and it seems to be handled perfectly on my side for weeks now.

**Comment by Monix362** [comment]
Image

I still get a readyState error.

**Comment by legendhimslef** [comment]
> Image
> 
> I still get a readyState error.

I have already fixed this issue, Please update your code

**Comment by Monix362** [comment]
Still the bot can not connect to the gateway, giving a lot of unavailable guilds and reconnects. 

Even if some shards have successfully connected, but there are 60-70 guilds left out of 1000 available guilds. So the bot will simply take a long time to connect to the gateway, leading to a huge amount of downtime.


Again, this happens when the bot has GUILD_PRESENCE intent... I didn't use it before, but now it is necessary for correct functioning of some bot services.

idea64_GjhYuBcR6N
ShareX_AMw89NjJsr

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by BenPatersonxx** [comment]
@legendhimslef Since your pull request got approved for v13.7, could you finish this PR?

**Review by iCrawl — APPROVED** [review]

**Pull request merged**